File: 314-binary-tree-vertical-order-traversal/314-binary-tree-vertical-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) return res;
        int left = 0, right = 0;
        //bfs
        Map<Integer, List<Integer>> verticalOrder = new HashMap<>();
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root, 0));
        while(!q.isEmpty()) {
            int size = q.size();
            for(int i =0; i < size; i++) {
                Pair node = q.poll();
                TreeNode curr = node.node;
                int currPos = node.pos;
                if(!verticalOrder.containsKey(currPos)) 
                    verticalOrder.put(currPos, new ArrayList<>());
                verticalOrder.get(currPos).add(curr.val);
                if(curr.left != null) q.offer(new Pair(curr.left, currPos - 1));
                if(curr.right != null) q.offer(new Pair(curr.right, currPos + 1));
                left = Math.min(left, currPos - 1);
                right = Math.max(right, currPos + 1);
            }
        }
        
        for(int i = left; i <= right; i++) {
            List<Integer> lst = verticalOrder.get(i);
            if(lst != null) res.add(lst);
        }
        return res;
    }
    class Pair {
        TreeNode node;
        int pos;
        public Pair(TreeNode node, int pos) {
            this.node = node;
            this.pos = pos;
        }
    }
}

File: 314-binary-tree-vertical-order-traversal/314-binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        if not root:
            return res
        q = deque([(root, 0)])
        left = right = 0
        vertical_map = defaultdict(list)
        
        while q:
            size = len(q)
            for i in range(size):
                node, vert_idx = q.popleft()
                vertical_map[vert_idx].append(node)
                left = min(left, vert_idx)
                right = max(right, vert_idx)
                if node.left:
                    q.append((node.left, vert_idx - 1))
                if node.right:
                    q.append((node.right, vert_idx +1))
        for i in range(left, right + 1):
            lst = vertical_map[i]
            if lst:
                res.append([node.val for node in lst])
        return res
                    